docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,498 @@
+    `_` and otherwise `UP(T0, T1)`
+- The expression `e0..e1` is inferred as `m0..m1` of type `T` in context `K`
+  - If `e0` is inferred as `m0` of type `T` in context `K`
+  - And `e1` is inferred as `m1` of type `P` in context `_`
+- The expression `e0 ? e1 : e2` is inferred as `m0 ? m1 : m2` of type `T` in
+  context `K`
+  - If `e0` is inferred as `m0` of any type in context `bool`
+  - And `e1` is inferred as `m1` of type `T0` in context `K`
+  - And `e2` is inferred as `m2` of type `T1` in context `K`
+  - Where `T` is the greatest closure of `K` with respect to `?` if `K` is not
+    `_` and otherwise `UP(T0, T1)`
+- TODO(leafp): Generalize the following closure cases to the full function
+  signature.
+  - In general, if the function signature is compatible with the context type,
+    take any available information from the context.  If the function signature
+    is not compatible, this this should always be a type error anyway, so
+    implementations should be free to choose the best error recovery path.
+  - The monomorphic case can be treated as a degenerate case of the polymorphic
+    rule
+- The expression `<T>(P x) => e` is inferred as `<T>(P x) => m` of type `<T>(P)
+  -> M` in context `_`
+  - If `e` is inferred as `m` of type `M` in context `_`
+- The expression `<T>(P x) => e` is inferred as `<T>(P x) => m` of type `<T>(P)
+  -> M` in context `<S>(Q) -> N`
+  - If `e` is inferred as `m` of type `M` in context `N[T/S]`
+  - Note: `x` is resolved as having type `P`  for inference in `e`
+- The expression `<T>(x) => e` is inferred as `<T>(dynamic x) => m` of type
+  `<T>(dynamic) -> M` in context `_`
+  - If `e` is inferred as `m` of type `M` in context `_`
+- The expression `<T>(x) => e` is inferred as `<T>(Q[T/S] x) => m` of type
+  `<T>(Q[T/S]) -> M` in context `<S>(Q) -> N`
+  - If `e` is inferred as `m` of type `M` in context `N[T/S]`
+  - Note: `x` is resolved as having type `Q[T/S]` for inference in `e`
+- Block bodied lamdas are treated essentially the same as expression bodied
+  lambdas above, except that:
+  - The final inferred return type is `UP(T0, ..., Tn)`, where the `Ti` are the
+    inferred types of the return expressions (`void` if no returns).
+  - The returned expression from each `return` in the body of the lamda uses the
+    same type expectation context as described above.
+  - TODO(leafp): flesh this out.
+- For async and generator functions, the downwards context type is computed as
+  above, except that the propagated downwards type is taken from the type
+  argument to the `Future` or `Iterable` or `Stream` constructor as appropriate.
+  If the return type is not the appropriate constructor type for the function,
+  then the downwards context is empty.  Note that subtypes of these types are
+  not considered (this is a strong mode error).
+- The expression `e(e0, .., ek)` is inferred as `m<M0, ..., Mn>(m0, ..., mk)` of
+  type `N` in context `_`:
```

Comment:
Are we missing a case here?  What if `e(e0, .., ek)` is inferred in a context other than `_`?